File: solutions/901/main.cc
#include <stack> 
using namespace std; 

class StockSpanner {
    stack<int> cnt_stack_; 
    stack<int> num_stack_; 
public:
    StockSpanner() {

    }
    
    int next(int price) {
        if (cnt_stack_.empty()) {
            cnt_stack_.push(1); 
            num_stack_.push(price); 
            return 1; 
        }

        // add today's count 
        int res = 1; 
        while (!num_stack_.empty() && price >= num_stack_.top()) {
            res += cnt_stack_.top(); 
            cnt_stack_.pop(); 
            num_stack_.pop(); 
        }
        cnt_stack_.push(res); 
        num_stack_.push(price);

        return res; 
    }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */

File: solutions/901/main.rs
use std::collections::VecDeque; 

struct StockSpanner {
    num_stack: VecDeque<i32>, 
    cnt_stack: VecDeque<i32>, 
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl StockSpanner {

    fn new() -> Self {
        Self {
            num_stack: VecDeque::new(), 
            cnt_stack: VecDeque::new(), 
        }
    }
    
    fn next(&mut self, price: i32) -> i32 {
        let mut res = 1; 
        if self.num_stack.len() == 0 {
            self.num_stack.push_back(price);
            self.cnt_stack.push_back(res); 
            return res; 
        }
        while self.num_stack.len() > 0 && price >= *self.num_stack.back().unwrap() {
            let cnt = self.cnt_stack.pop_back().unwrap(); 
            res += cnt; 
            self.num_stack.pop_back(); 
        }
        self.cnt_stack.push_back(res); 
        self.num_stack.push_back(price); 
        res 
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * let obj = StockSpanner::new();
 * let ret_1: i32 = obj.next(price);
 */